File: 297-serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        String s="";
        Queue<TreeNode> q=new LinkedList<>();
        if(root==null) return "";
        q.add(root);
        TreeNode ptr=root;

        while(!q.isEmpty()){
            TreeNode temp=q.remove();
            if(temp==null){
                s=s+"#"+" ";
                continue;
            }
            s=s+Integer.toString(temp.val)+" ";
            q.add(temp.left);
            q.add(temp.right);
        }
       
        System.out.println(s);
        return s;
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data.length()==0) return null;
        String[] values=data.split(" ");
        TreeNode root=new TreeNode(Integer.parseInt(values[0]));
        int n=values.length;
        Queue<TreeNode> q=new LinkedList<>();
        q.add(root);

        for(int i=1;i<n;i++){
            TreeNode ptr=q.remove();
            if(!values[i].equals("#")){
                ptr.left=new TreeNode(Integer.parseInt(values[i]));
                q.add(ptr.left);
            }
            if(!values[++i].equals("#")){
                ptr.right=new TreeNode(Integer.parseInt(values[i]));
                q.add(ptr.right);
            }
        }
        
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: 297-serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """

        q = []
        q.append(root)
        ans = ""

        while len(q) != 0:
            s = len(q)

            for i in range(s):
                node = q.pop(0)
                if node == None:
                    k = "#"
                else:
                    k = str(node.val)
                ans = ans + k + " "

                if node != None:
                    q.append(node.left)
                    q.append(node.right)
        
        return ans
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        
        q = []
        data = data[:len(data) - 1]
        dataa = data.split(" ")
        print(dataa)
        if len(dataa) == 1 and dataa[0] == '#':
            return None
        
        root = TreeNode(dataa[0])
        c = 1
        q.append(root)

        while len(q) != 0 and c < len(dataa):
            node = q.pop(0)

            if dataa[c] != '#':
                node.left = TreeNode(dataa[c])
                q.append(node.left)
            c += 1
            
            if c < len(data) and dataa[c] != '#':
                node.right = TreeNode(dataa[c])
                q.append(node.right)
            c += 1

        return root

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))